src/iotHubMessageExplorer.ts

```diff
@@ -21,27 +69,97 @@ export class IoTHubMessageExplorer extends IoTHubMessageBaseExplorer {
                 }
             }
         });
     }
 
+    private async sendD2CMessageCore(client: Client, message: string) {
+        let stringify = Utility.getConfig<boolean>(Constants.IoTHubD2CMessageStringifyKey);
+        client.sendEvent(new Message(stringify ? JSON.stringify(message) : message),
+            this.sendEventDone(client, Constants.IoTHubMessageLabel, Constants.IoTHub, Constants.IoTHubAIMessageDoneEvent));
+    }
+
+    private async sendD2CMessageCoreWithProgress(client: Client, message: string, status: SendStatus, totalStatus: SendStatus) {
+        let stringify = Utility.getConfig<boolean>(Constants.IoTHubD2CMessageStringifyKey);
+        await client.sendEvent(new Message(stringify ? JSON.stringify(message) : message),
+            this.sendEventDoneWithProgress(client, Constants.IoTHubAIMessageDoneEvent, status, totalStatus));
+    }
+
+    private async delay(ms: number, token?: vscode.CancellationToken) {
+        if (ms <= 1000) {
+            return new Promise( resolve => setTimeout(resolve, ms));
+        } else {
+            await new Promise( resolve => setTimeout(resolve, 1000));
+            if (token) {
+                if (token.isCancellationRequested) {
+                    return;
+                } else {
+                    await this.delay(ms - 1000, token);
+                }
+            } else {
+                await this.delay(ms - 1000);
+            }
+        }
+    }
+
+    public async sendD2CMessageFromMultipleDevicesRepeatedlyWithProgressBar(deviceConnectionStrings: string[], message: string, times: number, interval: number) {
+        const startTime = new Date();
+        this.outputLine(Constants.SimulatorSummaryLabel, `Start at ${startTime}`)
+        await vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification,
+			title: Constants.SimulatorSendingMessageProgressBarTitle,
+			cancellable: true
+		}, async (progress, token) => {
+            token.onCancellationRequested(() => {
+                vscode.window.showInformationMessage(Constants.SimulatorProgressBarCancelLog);
+            })
+            progress.report({ increment: 0});
+            const deviceCount = deviceConnectionStrings.length;
+            const total = deviceCount * times;
+            const step = 100 / total;
+            let count = 0;
+            let clients = [];
+            let statuses = [];
+            let totalStatus = new SendStatus('Total', total);
+            for(let i = 0; i < deviceCount; i++) {
+                clients.push(await clientFromConnectionString(deviceConnectionStrings[i]));
+                statuses.push(new SendStatus(ConnectionString.parse(deviceConnectionStrings[i]).DeviceId, times));
+            }
+            for(let i = 0; i < times; i++) {
+                for (let j = 0; j < deviceCount; j++) {
+                    this.sendD2CMessageCoreWithProgress(clients[j], message, statuses[j], totalStatus);

```

Comment:
No await here? Seems need await Promise.all in this scenario?

Reply:
```suggestion
                    await ids.map(async j => await this.sendD2CMessageCoreWithProgress(clients[j], message, statuses[j], totalStatus));
```

Reply:
Modified.